[#67622] Select hierarchy items with filterable tree view

Kharonus opened the pull request:
# Ticket
[#67622](https://community.openproject.org/work_packages/67622)

# What are you trying to accomplish?
- use filterable tree view to select values for project attributes

# What approach did you choose and why?
- render filterable tree view in edit dialog for project attributes
- preselect current values
- assign new values

Kharonus committed:
[#67622] Select hierarchy items with filterable tree view

- https://community.openproject.org/work_packages/67622
- render filterable tree view in edit dialog for project attributes
- preselect current values
- assign new values

Kharonus committed:
[#67622] use render_tree helper for change parent and overview

Kharonus committed:
[#67622] readded current state for overview tree

NobodysNightmare commented:
I found an issue, where I am not sure how it happens and whether it's a bug of this PR or a separate problem.

## Reproduction

1. Have a project attribute of type hierarchy or scored list
2. Fill out a value already and save it
3. Click on the value, to bring up the edit modal
4. Immediately click "Save"

## Expected behaviour

It closes the modal, effectively without changes, because I am submitting the same state that I had before.

## Actual behaviour

The modal remains open and nothing happens. Inspecting the network tab reveals, that the network request fails with an HTTP 400.

Performing a no-op change in the treeview fixes the issue. For example clicking on the selected item in a single select or selecting and deselecting an item in a multi-select. Afterwards the tree view is submittable.

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/change_item_parent_dialog_component.html.erb:
```diff
@@ -44,7 +44,19 @@ See COPYRIGHT and LICENSE files for more details.
-          add_sub_tree(tree_view, hierarchy_items)
+          current_fn = lambda { |item| item.id == @hierarchy_item.id }
+          disabled_fn = lambda { |item| current_fn.call(item) || item.id == @hierarchy_item.parent.id }
+          expanded_fn = lambda { |item| hierarchy_service.descendant_of?(item: @hierarchy_item, parent: item).success? }
```
:yellow_circle: I recognize that `descendant_of?` was implemented outside the scope of this PR, but I was confused anyways to read it as `descendant_of?(...).success?`

What was the motivation behind that? IMO it looks like an abusive use of dry monads:

**Problem 1**: The `?` at the end of `descendant_of?` gives me the impression that the result of the method is usable as a truthy/falsey value, which is it not, because the result is always truthy:

```ruby
pry(main)> Failure ? "truthy" : "falsey"
=> "truthy"
pry(main)> Success ? "truthy" : "falsey"
=> "truthy"
```

**Problem 2**: `Success` and `Failure` would usually indicate whether something succeeded or failed, but they are misused here. I asked the question "is X a descendant of Y" and I successfully get an answer all the time. And sometimes the answer is no. That's not a failure. It's one of two answers the caller was expecting when asking the question.

I am really surprised about this implementation. So surprised, that I bring it up in this PR, which is just using this method.

Kharonus replied:
IMHO this implementation is not so surprising, once you view the whole context of the service. The service responds always with a result monad. And when you think about how to wrap a boolean into a monadic response, then the answer is "you don't" the result itself is already carrying a boolean expression.

Maybe the confusion stems from the perspective, that a "Failure" is something that goes wrong in code. But instead a failure can simply be "No" answer.

NobodysNightmare replied:
I find it fascinating that we are discussing the "surprise" aspect here. You have a method that's called `blubb?` and the fact that `if blubb?` will always go into the `if` branch, does not qualify as a surprise for you?

> Maybe the confusion stems from the perspective, that a "Failure" is something that goes wrong in code

I wouldn't call it "goes wrong in code" (which sounds like a programming error), but yes: I'd only expect `Failure` to be returned in **error** cases. Which this is not.

Let me [quote the authors](https://dry-rb.org/gems/dry-monads/1.0/result/) here:

> The `Result` mixin has two type constructors: `Success` and `Failure`. The `Success` can be thought of as "everything went success" and the `Failure` is used when "something has gone wrong".

And this is exactly my point: Nothing has gone wrong, if `a` is not a descendant of `b`.

NobodysNightmare replied:
To me this looks like a classic "when all you have is a hammer, everything looks like a nail". Just because we got exactly two result values and just because booleans also happen to be exactly two values, doesn't mean that results are a good replacement for booleans as a general concept. If the boolean is called `success`, then a result is probably a good replacement, but otherwise not.

NobodysNightmare commented on app/helpers/custom_field_hierarchy_tree_view_helper.rb:
```diff
@@ -0,0 +1,72 @@
+
+    return {} if hierarchy_hash.nil?
+
+    if options[:show_root]
```
:yellow_circle: As discussed earlier, `options` is a pretty heavy parameter. It would probably make sense to focus it on a specific job and name it accordingly.

What we discussed was to extract `:show_root` into a keyword argument to `render_tree` and then rename options to something like `item_options`, indicating that the options are all passed on to the rendering of items.

NobodysNightmare commented on app/helpers/custom_field_hierarchy_tree_view_helper.rb:
```diff
@@ -0,0 +1,72 @@
+#++
+
+module CustomFieldHierarchyTreeViewHelper
+  def render_tree(custom_field, tree_view, options = {})
```
:green_circle: I am wondering whether it would help the readability if we swapped the arguments here:

```suggestion
  def render_tree(tree_view, custom_field, options = {})
```

This would be consistent with the `add_sub_tree` method, which also receives the tree (the thing that we add something to) as a first argument.

NobodysNightmare replied:
:green_circle: Another thought regarding this method (that came to me after comparing it to `add_sub_tree`):

What this method seems to be doing is to add the values of the custom field to the tree view. That the tree view is empty at the beginning of the process is nice, but I guess the method would also work nicely if that wasn't the case. Also the method is not rendering the tree, but just adding items to it.

Should we rename the method?

* `add_to_tree_view`
* `populate_tree_view`

NobodysNightmare commented on app/helpers/custom_field_hierarchy_tree_view_helper.rb:
```diff
@@ -0,0 +1,72 @@
+  def render_tree(custom_field, tree_view, options = {})
+    hierarchy_hash = custom_field.hierarchy_root.hash_tree
+
+    return {} if hierarchy_hash.nil?
```
:yellow_circle: Is the return value of this method relevant at all? Since it's always at the end of the `render` block for the tree view, it seems plausible. But on the other hand we'd return the result of `add_sub_tree` in the normal case, which would be just one random `each` result. To me it seems that the side effect of calling `add_sub_tree` is the important thing, because that calls methods on `tree_view`.

My current hypothesis is that the return value plays no role and we could simplify this to

```suggestion
    return if hierarchy_hash.nil?
```

Kharonus replied:
yeah, this is a copy paste error, good eye

NobodysNightmare commented on modules/overviews/app/components/overviews/project_custom_fields/edit_component.html.erb:
```diff
@@ -8,14 +8,45 @@
+          )
+        ) do |tree_view|
+          current_value = @project.custom_value_for(@project_custom_field)
+          checked_fn = lambda do |item|
```
:yellow_circle: Because you were not satisfied with the need for a `case` statement in this lambda. I think you don't need this, if you always think in terms of arrays:

```ruby
# Array(...) wraps non-array, but keeps arrays, so afterwards we have a CustomValue[]
current_values = Array(@project.custom_value_for(@project_custom_field)).map(&:value)
checked_fn = ->(item) { current_values.include?(item.id.to_s) }
```

Kharonus replied:
Oh, okay. Interesting workaround, but I like

NobodysNightmare commented on modules/overviews/app/controllers/overviews/project_custom_fields_controller.rb:
```diff
@@ -64,6 +70,18 @@ def update
+      MultiJson.load(value, symbolize_keys: true)[:value]
+    end
+
+    params[:project][:custom_field_values] = { @custom_field.id.to_s => ids.one? ? ids.first : ids }
```
:red_circle: Not sure at which point it fails (I thought about this edge case when reading this line, so I am commenting here): If you try to select nothing in a multi-select hierarchy, the request fails. For single select fields, there is not even a UI to select nothing.

Is that intended? My fields are not required and they started without a value. So I'd expect that I'll be able to bring them back to "no value" as well.

Kharonus replied:
Very good point. for single select we NEED a remove. And that multi select fails is an oversight on my side.

Kharonus replied:
Even deselection in multiselect boils down to a tree view error, I think. The form is submitted without the form field parameter - which should have an empty value if nothing is selected, but it is not existent.

Kharonus replied:
I've solved the deselection for multi select variants, and the single select will be handled by the tree view no in https://community.openproject.org/wp/68697

Kharonus committed:
[#67622] handled code review improvements

- rename render_tree to populate_tree_view to improve understanding of
  the method's purpose
- refined the methods arguments

Kharonus committed:
[#67622] enable deselection of hierarchy project attributes
